File: neetcode/medium/remove-node-from-end-of-linked-list.py
from operator import index
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        first = head
        cur = head
        length = 0
        while cur:
            length += 1
            cur = cur.next

        index_to_remove = length - n
        cur_i = 0
        cur = first
        if index_to_remove == 0:
            return first.next
        while cur_i <= index_to_remove:
            if cur_i == index_to_remove - 1:
                cur.next = cur.next.next
                break
            cur = cur.next
            cur_i += 1

        return first


def create_list(arr):
    if not arr:
        return None
    head = ListNode(arr[0])
    current = head
    for i in range(1, len(arr)):
        current.next = ListNode(arr[i])
        current = current.next
    return head


list2 = create_list([1, 2, 3, 4])

Solution().removeNthFromEnd(list2, 2)

File: neetcode/medium/remove-node-from-end-of-linked-list.js
// /**
class ListNode {
  constructor(val = 0, next = null) {
    this.val = val;
    this.next = next;
  }
}

class Solution {
  /**
   * @param {ListNode} head
   * @param {number} n
   * @return {ListNode}
   */
  removeNthFromEnd(head, n) {
    let first = head;
    let length = 0;
    let cur = head;
    while (cur) {
      length += 1;
      cur = cur.next;
    }
    const indexToRemove = length - n;
    cur = first;
    if (indexToRemove === 0) {
      return first.next;
    }
    for (let i = 0; i < length; i++) {
      if (i === indexToRemove - 1) {
        cur.next = cur.next.next;
        break;
      }
      cur = cur.next;
    }
    return first;
  }
}

// Test case: [1,2,3,4], n = 2
const head = new ListNode(1);
head.next = new ListNode(2);
head.next.next = new ListNode(3);
head.next.next.next = new ListNode(4);

const n = 2;
const solution = new Solution();
const result = solution.removeNthFromEnd(head, n);

let values = [];
let node = result;
while (node) {
  values.push(node.val);
  node = node.next;
}
console.log(values);
